src/test/java/com/browserstack/test/suites/TestBase.java

```diff
@@ -0,0 +1,121 @@
+        String url = (String) config.get("application_endpoint");
+        if (environment.equalsIgnoreCase("local")) {
+            if (OsUtils.isMac()) {
+                System.setProperty("webdriver.chrome.driver", Path.of(driverLocation, "/chromedriver").toString());
+            }
+            if (OsUtils.isWindows()) {
+                System.setProperty("webdriver.chrome.driver", Path.of(driverLocation, "/chromedriver.exe").toString());
+            }
+            if (OsUtils.isUnix()) {
+                System.setProperty("webdriver.chrome.driver", Path.of(driverLocation, "/chromedriver").toString());
+            }
+            driver.set(new ChromeDriver());
+        } else if (environment.equalsIgnoreCase("remote")) {
+            JSONObject profilesJson = (JSONObject) config.get("tests");
+            JSONObject envs = (JSONObject) profilesJson.get(test);
+
+            Map<String, String> commonCapabilities = (Map<String, String>) envs.get("common_caps");
+            commonCapabilities.put("name",m.getName());
+            Map<String, String> envCapabilities = (Map<String, String>) ((org.json.simple.JSONArray) envs.get("env_caps")).get(env_cap_id);
+            Map<String, String> localCapabilities = (Map<String, String>) envs.get("local_binding_caps");
+
+            DesiredCapabilities caps = new DesiredCapabilities();
+            caps.merge(new DesiredCapabilities(commonCapabilities));
+            caps.merge(new DesiredCapabilities(envCapabilities));
+            if (test.equals("local")) {
+                url = (String) envs.get("application_endpoint");
+                caps.merge(new DesiredCapabilities(localCapabilities));
+            }
+
+            String username = System.getenv("BROWSERSTACK_USERNAME");
+            if (username == null) {
+                username = (String) config.get("user").toString();
+            }
+            String accessKey = System.getenv("BROWSERSTACK_ACCESS_KEY");
+            if (accessKey == null) {
+                accessKey = (String) config.get("key").toString();
+            }
+
+            if (!getGeolocation().isEmpty()) {
+                caps.setCapability("browserstack.geoLocation", getGeolocation());
+            }
+
+            if (caps.getCapability("browserstack.local") != null
+                    && caps.getCapability("browserstack.local").equals("true")) {
+                local = new Local();
+                UUID uuid = UUID.randomUUID();
+                caps.setCapability("browserstack.localIdentifier", uuid.toString());
+                Map<String, String> options = new HashMap<String, String>();
+                options.put("key", accessKey);
+                options.put("localIdentifier", uuid.toString());
+                local.start(options);
+            }
+
+            driver.set(new RemoteWebDriver(new URL("https://" + username + ":" + accessKey + "@hub.browserstack.com/wd/hub"), caps));
```

Comment:
use class level variables for HUB URL.
example: 
https://github.com/browserstack/browserstack-examples-cucumber-junit5/blob/review/src/test/java/com/browserstack/steps/SetupSteps.java#L38

Also use capabilities for user and access key instead of passing it in the URL. The creds stay safe from MITM that way.


browserstack.userFor running your Selenium and Appium tests on BrowserStack, it requires a username and an access key for authenticating the user. This capability can be used to set the username. | You can find your username and access key on the Settings page under the Automate section.Example: "browserstack.user": "<BROWSERSTACK_USERNAME>"
-- | --
browserstack.keyFor running your Selenium and Appium tests on BrowserStack, it requires a username and an access key for authenticating the user. This capability can be used to set the access key. | You can find your username and access key on the Settings page under the Automate section.Example: "browserstack.key": "<BROWSERSTACK_ACCESSKEY>"